src/enzyme_ad/jax/Passes/CanonicalizeFor.cpp

```diff
@@ -2738,27 +2738,276 @@ struct SelectI1Simplify : public OpRewritePattern<arith::SelectOp> {
   }
 };
 
+void checkOperands(
+    scf::IfOp ifOp, int opOperandIndex, Value operandIf, Value operandElse,
+    SmallVector<std::tuple<unsigned long, Operation *, unsigned long>>
+        &yieldUseMap,
+    SmallVector<Operation *> &opsToMoveAfterIf,
+    SmallVector<Value> &ifYieldOperands, SmallVector<Value> &elseYieldOperands,
+    PatternRewriter &rewriter) {
+  // Extract unique users of the operands and checking if count is 1
+  llvm::SmallPtrSet<Operation *, 4> uniqueThenUsers, uniqueElseUsers;
+  for (Operation *user : operandIf.getUsers()) {
+    uniqueThenUsers.insert(user);
+  }
+  for (Operation *user : operandElse.getUsers()) {
+    uniqueElseUsers.insert(user);
+  }
+  if (isa<BlockArgument>(operandIf) || isa<BlockArgument>(operandElse) ||
+      !ifOp.getThenRegion().isAncestor(
+          operandIf.getDefiningOp()->getParentRegion()) ||
+      !ifOp.getElseRegion().isAncestor(
+          operandElse.getDefiningOp()->getParentRegion()) ||
+      !(uniqueThenUsers.size() == 1 && uniqueElseUsers.size() == 1) ||
+      (operandIf.getDefiningOp()->getName() !=
+       operandElse.getDefiningOp()->getName()) ||
+      !isReadNone(operandIf.getDefiningOp()) ||
+      !isReadNone(operandElse.getDefiningOp())) {
+    ifYieldOperands.push_back(operandIf);
+    elseYieldOperands.push_back(operandElse);
+    yieldUseMap.push_back(std::make_tuple(
+        ifYieldOperands.size() - 1, opsToMoveAfterIf.back(), opOperandIndex));
+    return;
+  }
+
+  Operation *opToMove = operandIf.getDefiningOp();
+  opsToMoveAfterIf.push_back(opToMove);
+  int index = 0;
```

Comment:
you can also use llvm::enumerate to remove this variable. I left an example below.